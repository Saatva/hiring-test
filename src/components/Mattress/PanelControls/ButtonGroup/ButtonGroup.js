import React, { useState, useRef, useEffect } from 'react';

import buttonGroupClasses from './ButtonGroup.module.scss';

const ButtonGroup = (props) => {
    const [clickedBtnIndex, setClickedBtnIndex] = useState(0);
    const btnGroupElm = useRef(null);
    const btnIndxRef = useRef(0);
    const buttonsRef = useRef([]);

    const clickHandler = (mattress, i) => {
        props.buttonClicked(mattress);
        setClickedBtnIndex(i)
        btnIndxRef.current = i;
    };

    const buildBtnGroup = (mattresses) => {
        return (
            mattresses.map((mattress, i) => (
                <button
                    key={i}
                    ref={el => buttonsRef.current[i] = el}
                    name={mattress.name}
                    onClick={() => clickHandler(mattress, i)}
                >
                    {mattress.name}
                </button>
            ))
        )
    };

    const handleClickCallback = e => {
        if (btnGroupElm.current.contains(e.target)) {
            return;
        }

        if (clickedBtnIndex === btnIndxRef.current)
            setClickedBtnIndex(-1)
        else
            setClickedBtnIndex(clickedBtnIndex)
    };

    useEffect(() => {
        buttonsRef.current[btnIndxRef.current].focus();
        setClickedBtnIndex(btnIndxRef.current)
    }, [clickedBtnIndex]);

    useEffect(() => {
        buttonsRef.current[0].focus();
        document.addEventListener("mousedown", handleClickCallback);

        return () => {
            document.removeEventListener("mousedown", handleClickCallback);
        };
    }, []);

    return (
        <div className={buttonGroupClasses.ButtonGroup} ref={btnGroupElm}>
            {buildBtnGroup(props.mattresses)}
        </div>
    )
};

export default ButtonGroup;
