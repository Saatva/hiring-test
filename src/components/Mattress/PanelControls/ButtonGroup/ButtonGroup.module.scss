.ButtonGroup {
  display: flex;
  width: 100%;
  button {
    padding: 10px 24px;
    cursor: pointer;
    float: left;
    font-family: 'Questrial', sans-serif;
    font-size: 1rem;
    font-weight: 400;
    line-height: 1;
    outline: none;
    color: #6b6257;
    background-color: #fff;
    border: 1px solid #d2d2d2;
    text-align: center;
    white-space: nowrap;
    flex: 1;
    &:not(:last-child) {
      border-right: none;
    }
    &:hover {
      background-color: hsl(0, 0%, 95%);
    }
    &:focus,
    &:active {
      color: #fff;
      background-color: #a6a19a;
      border-color: #d2d2d2;
    }
  }
}
