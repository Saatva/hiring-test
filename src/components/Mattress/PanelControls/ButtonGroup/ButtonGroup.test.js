import React from 'react';
import ButtonGroup from './ButtonGroup';
import { configure, shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

configure({ adapter: new Adapter() });

describe('<ButtonGroup />', () => {
    const buttonClicked = jest.fn();

    const container = shallow(<ButtonGroup buttonClicked={buttonClicked} mattresses={[{name: 'Classic', price: 120}, {name: 'Loom', price: 220}]} />);

    it('should have a button for each mattress', () => {
        expect(container.find('button').length).toEqual(2);
    });

    it('should call buttonClicked handler with the correct mattress when button is clicked', () => {
        const handleClick = jest.spyOn(React, "useState");
        handleClick.mockImplementation(mattress => [mattress, buttonClicked]);

        container.find('button').at(1).simulate('click');

        expect(buttonClicked).toHaveBeenCalledWith({name: 'Loom', price: 220})
    });

});