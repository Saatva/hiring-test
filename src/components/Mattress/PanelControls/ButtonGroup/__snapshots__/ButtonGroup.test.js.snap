// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup /> should match the snapshot 1`] = `"<div class=\\"ButtonGroup\\"><button name=\\"Classic\\">Classic</button></div>"`;

exports[`<ButtonGroup /> should match the snapshot 2`] = `"<div class=\\"ButtonGroup\\"><button name=\\"Classic\\">Classic</button><button name=\\"Loom\\">Loom</button></div>"`;
